modules/axtask/src/wait_queue.rs

```diff
@@ -27,49 +29,54 @@ use crate::{AxRunQueue, AxTaskRef, CurrentTask, RUN_QUEUE};
 /// assert_eq!(VALUE.load(Ordering::Relaxed), 1);
 /// ```
 pub struct WaitQueue {
-    queue: SpinRaw<VecDeque<AxTaskRef>>, // we already disabled IRQs when lock the `RUN_QUEUE`
+    queue: SpinNoIrq<VecDeque<AxTaskRef>>,
 }
 
+pub(crate) type WaitQueueGuard<'a> = SpinNoIrqGuard<'a, VecDeque<AxTaskRef>>;
+
 impl WaitQueue {
     /// Creates an empty wait queue.
     pub const fn new() -> Self {
         Self {
-            queue: SpinRaw::new(VecDeque::new()),
+            queue: SpinNoIrq::new(VecDeque::new()),
         }
     }
 
     /// Creates an empty wait queue with space for at least `capacity` elements.
     pub fn with_capacity(capacity: usize) -> Self {
         Self {
-            queue: SpinRaw::new(VecDeque::with_capacity(capacity)),
+            queue: SpinNoIrq::new(VecDeque::with_capacity(capacity)),
         }
     }
 
-    fn cancel_events(&self, curr: CurrentTask) {
+    /// Cancel events by removing the task from the wait queue.
+    /// If `from_timer_list` is true, the task should be removed from the timer list.
+    fn cancel_events(&self, curr: CurrentTask, from_timer_list: bool) {
         // A task can be wake up only one events (timer or `notify()`), remove
         // the event from another queue.
         if curr.in_wait_queue() {
             // wake up by timer (timeout).
-            // `RUN_QUEUE` is not locked here, so disable IRQs.
-            let _guard = kernel_guard::IrqSave::new();
-            self.queue.lock().retain(|t| !curr.ptr_eq(t));
+            let mut wq_locked = self.queue.lock();
+            wq_locked.retain(|t| !curr.ptr_eq(t));
             curr.set_in_wait_queue(false);
         }
+        // Try to cancel a timer event from timer lists.
+        // Just mark task's current timer ticket ID as expired.
         #[cfg(feature = "irq")]
-        if curr.in_timer_list() {
-            // timeout was set but not triggered (wake up by `WaitQueue::notify()`)
-            crate::timers::cancel_alarm(curr.as_task_ref());
+        if from_timer_list {
+            curr.timer_ticket_expired();
+            // TODO:
+            // this task is still not removed from timer list of target CPU,
```

Comment:
I DO NOT think so, we just mark this timer event as expired, instead of removing it from timer list, it will still trigger a timer event, enter `callback` and be ignored due to Timer ticket ID not matched.

Reply:
It can be considered a lazy-removal strategy, ignoring it when it is about to take effect.

Reply:
However, it still hasn't been completely removed and needs to go through the process of expiring an event from the timer list and invoking the callback.

Reply:
I mean delete `TODO`, there is nothing to do.

Reply:
Ok, the comments have been updated